Wrappers/Python/cil/optimisation/algorithms/SPDHG.py

```diff
@@ -41,203 +45,422 @@ class SPDHG(Algorithm):
+
+        .. math:: \tau = 0.99\min_i( \frac{p_i}{ (\sigma_i  \|K_i\|^2) })
+
+        - Case 3: If `tau` is provided but not `sigma` then `sigma` is calculated using the formula
+
+        .. math:: \sigma_i= \frac{0.99 p_i}{\tau\|K_i\|^2}
+
+        - Case 4: Both `sigma` and `tau` are provided.
+
+
+        Parameters
+        ----------
+            sigma : list of positive float, optional, default= see docstring
+                List of Step size parameters for dual problem
+            tau : positive float, optional, default= see docstring
+                Step size parameter for primal problem
+
+        """
+        gamma = 1.
+        rho = .99
+        if sigma is not None:
+            if len(sigma) == self._ndual_subsets:
+                if all(isinstance(x, Number) and x > 0 for x in sigma):
+                    pass
+                else:
+                    raise ValueError(
+                        "Sigma expected to be a positive number.")
+
+            else:
+                raise ValueError(
+                    "Please pass a list of floats to sigma with the same number of entries as number of operators")
+            self._sigma = sigma
+
+        elif tau is None:
+            self._sigma = [gamma * rho / ni for ni in self._norms]
+        else:
+            self._sigma = [
+                rho*pi / (tau*ni**2) for ni, pi in zip(self._norms, self._prob_weights)]
+
+        if tau is None:
+            values = [rho*pi / (si * ni**2) for pi, ni,
+                      si in zip(self._prob_weights, self._norms, self._sigma)]
+            self._tau = min([value for value in values if value > 1e-8])
+
+        else:
+            if isinstance(tau, Number) and tau > 0:
+                pass
+            else:
+                raise ValueError(
+                    "The step-sizes of SPDHG must be positive, passed tau = {}".format(tau))
+
+            self._tau = tau
+
+    def check_convergence(self):
+        """  Checks whether convergence criterion for SPDHG is satisfied with the current scalar values of tau and sigma
+
+        Returns
+        -------
+        Boolean
+            True if convergence criterion is satisfied. False if not satisfied or convergence is unknown. 
+
+        Note
+        -----
+        Convergence criterion currently can only be checked for scalar values of tau.
+
+        Note
+        ----
+        This checks the convergence criterion. Numerical errors may mean some sigma and tau values that satisfy the convergence criterion may not converge. 
+        Alternatively, step sizes outside the convergence criterion may still allow (fast) convergence. 
+        """
+        for i in range(self._ndual_subsets):
+            if isinstance(self._tau, Number) and isinstance(self._sigma[i], Number):
+                if self._sigma[i] * self._tau * self._norms[i]**2 > self._prob_weights[i]:
+                    return False
+                return True
+            else:
+                return False
```

Comment:
do you need this?

Reply:
I changed it to return a value error instead. Basically we can't check it for non-scalar values of tau